api/import.go

```diff
@@ -52,3 +56,31 @@ func ImportChannel(channel *model.Channel) *model.Channel {
 		return sc
 	}
 }
+
+func ImportFile(file io.Reader, teamId string, channelId string, userId string, fileName string) (string, error) {
+	buf := bytes.NewBuffer(nil)
+	io.Copy(buf, file)
+
+	uid := model.NewId()
+
+	imageNameList := []string{}
+	imageDataList := [][]byte{}
+
+	if model.IsFileExtImage(filepath.Ext(fileName)) {
+		imageNameList = append(imageNameList, uid+"/"+fileName)
+		imageDataList = append(imageDataList, buf.Bytes())
+	}
+
+	path := "teams/" + teamId + "/channels/" + channelId + "/users/" + userId + "/" + uid + "/" + fileName
+
+	if err := WriteFile(buf.Bytes(), path); err != nil {
+		return "", err
+	}
+
+	encName := utils.UrlEncode(fileName)
+	fileUrl := "/" + channelId + "/" + userId + "/" + uid + "/" + encName
+
+	go handleImages(imageNameList, imageDataList, teamId, channelId, userId)
+
+	return fileUrl, nil
+}
```

Comment:
Ignoring what I said about each file needing a post, this will still conflict with #4068 since files will now need to be stored in the database before they can be attached to a post. If this gets merged in after my changes, a couple things will need to be changed here to properly save and attach the files.

There's 3 things that'll need to be changed to make this work with my stuff:
1. Create a `FileInfo` object for the file and save it to the DB. This should be easy since I added a
```
func doUploadFile(teamId string, channelId string, userId string, rawFilename string, data []byte) (*model.FileInfo, *model.AppError)
```
function with my changes that will do that along with writing the file to disk. You'll still need to call `handleImages` with its new arguments, but the rest of the stuff here will be done automatically.
2. Instead of saving the `fileUrl` into the `Post.Filenames` field, you'll save the `FileInfo`'s `Id` into the `Post.FileIds` field. That's basically the same as what you're already doing.
3. You'll need to change the `ImportPost` function above to attach the `FileInfos` back to the `Post` like I do in [the updated `CreatePost` function](https://github.com/mattermost/platform/blob/plt3105/api/post.go#L146). You might even be able to just copy and paste that block to do that.

And that should be it. We'll have to see who's PR goes in first though.